src/java/org/apache/cassandra/service/StorageService.java

```diff
@@ -3291,12 +3291,29 @@ public String getKeyspaceReplicationInfo(String keyspaceName)
         return stringify(Gossiper.instance.getUnreachableMembers(), true);
     }
 
+    @Override
     public String[] getAllDataFileLocations()
     {
-        String[] locations = DatabaseDescriptor.getAllDataFileLocations();
-        for (int i = 0; i < locations.length; i++)
-            locations[i] = FileUtils.getCanonicalPath(locations[i]);
-        return locations;
+        return getCanonicalPaths(DatabaseDescriptor.getAllDataFileLocations());
+    }
+
+    private String[] getCanonicalPaths(String[] paths)
+    {
+        for (int i = 0; i < paths.length; i++)
+            paths[i] = FileUtils.getCanonicalPath(paths[i]);
+        return paths;
+    }
```

Comment:
Overwriting the input array seems like it could have risky side effects, like overriding the contents of `DatabaseDescriptor.conf.data_file_directories`. I think we should return a new array.

Reply:
Good catch. :-)